draco/analysis/sourcestack.py

```diff
@@ -77,101 +83,118 @@ def process(self, formed_beam):
             self.stack_axis["centre"] - self.stack_axis["centre"][self.freqside]
         )
 
-        # Get f_mask and qs_indices
-        freqdiff = freq[np.newaxis, :] - qso_freq[:, np.newaxis]
-        # Stack axis bin edges to digitize each quasar at.
-        stackbins = self.stack_axis["centre"] + 0.5 * self.stack_axis["width"]
-        stackbins = np.append(
-            stackbins,
-            self.stack_axis["centre"][-1] - 0.5 * self.stack_axis["width"][-1],
-        )
-        # Index of each frequency in stack axis, for each quasar
-        qs_indices = np.digitize(freqdiff, stackbins) - 1
-        # Indices to be processed in full frequency axis for each quasar
-        f_mask = (qs_indices >= 0) & (qs_indices < self.nstack)
-        # Only quasars in the frequency range of the data.
-        qso_mask = (np.sum(f_mask, axis=1) > 0).astype(bool)
+        # Get f_mask and source_indices
+        freqdiff = freq[np.newaxis, :] - source_freq[:, np.newaxis]
+
+        # Stack axis bin edges to digitize each source at, in either increasing
+        # or decreasing order depending on order of frequencies
+        if self.stack_axis["centre"][0] > self.stack_axis["centre"][-1]:
+            stackbins = self.stack_axis["centre"] + 0.5 * self.stack_axis["width"]
+            stackbins = np.append(
+                stackbins,
+                self.stack_axis["centre"][-1] - 0.5 * self.stack_axis["width"][-1],
+            )
+        else:
+            stackbins = self.stack_axis["centre"] - 0.5 * self.stack_axis["width"]
+            stackbins = np.append(
+                stackbins,
+                self.stack_axis["centre"][-1] + 0.5 * self.stack_axis["width"][-1],
+            )
+        # Index of each frequency in stack axis, for each source
+        source_indices = np.digitize(freqdiff, stackbins) - 1
+        # Indices to be processed in full frequency axis for each source
+        f_mask = (source_indices >= 0) & (source_indices < self.nstack)
+        # Only sources in the frequency range of the data.
+        source_mask = (np.sum(f_mask, axis=1) > 0).astype(bool)
+        # If desired, also restrict to sources within a specific channel.
+        # This works because the frequency axis is not distributed between
+        # ranks.
+        if self.single_source_bin_index is not None:
+            restricted_chan_mask = (
```

Comment:
I think this expression will fail for edge bins, in the sense that all sources that fall outside the range of frequencies contained in the formed beam will be included in the edge bin's `restricted_chan_mask`.  I'm not sure if this is an issue for your use case.

I think something like:
```
fs = formed_beam.index_map['freq'][self.single_source_bin_index]
restricted_chan_mask = np.abs(source_freq - fs['centre']) < (0.5 * fs['width'])
```
would handle that particular edge case properly.